bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecProviderTest.kt

```diff
@@ -75,6 +79,41 @@ class KotlinSerializerCodecProviderTest {
         assertEquals(DataClassWithSimpleValues::class.java, codec.encoderClass)
     }
 
+    @Test
+    fun testDataClassWithSimpleValuesFieldOrdering() {
+        val codec = MongoClientSettings.getDefaultCodecRegistry().get(DataClassWithSimpleValues::class.java)
+        val expected = DataClassWithSimpleValues('c', 0, 1, 22, 42L, 4.0f, 4.2, true, "String")
+
+        val numberLong = "\$numberLong"
+        val actual =
+            codec.decode(
+                JsonReader(
+                    """{"boolean": true, "byte": 0, "char": "c", "double": 4.2, "float": 4.0, "int": 22,
+                        |"long": {"$numberLong": "42"}, "short": 1, "string": "String"}"""
+                        .trimMargin()),
+                DecoderContext.builder().build())
+
+        assertEquals(expected, actual)
+    }
+
+    @Test
+    fun testDataClassSealedFieldOrdering() {
+        val codec = MongoClientSettings.getDefaultCodecRegistry().get(SealedInterface::class.java)
+
+        val objectId = ObjectId("111111111111111111111111")
+        val oid = "\$oid"
+        val expected = DataClassSealedInterface(objectId, "string")
+        val actual =
+            codec.decode(
+                JsonReader(
+                    """{"name": "string", "_id": {$oid: "${objectId.toHexString()}"},
```

Comment:
Here all the fields are out of order and this confirms it does support the scenario where the discriminator isn't the first field.